README.md
# linllt

## What is linllt?

The linllt libraries (libmescan and libllt) are aravis-based libraries for operating scanCONTROL sensors from MICRO-EPSILON. libmescan is a C library for sensor control and data aquisition. linllt is a C++ wrapper.

## Building the linllt libraries

For pain-free usage, the libraries can be built with the [Meson](http://mesonbuild.com) build system. Meson checks for the required dependencies, like aravis, and does all necessary steps to prepare compiling and installing the code on the currently used platform. For compiling and installing [Ninja](https://ninja-build.org) is used.

If Meson is not installed on the host PC, it can be retrieved via the distribuitions package manager, e.g.

```bash
sudo apt install meson
```

or from [source](https://github.com/mesonbuild/meson/releases). Meson dependencies are [Python](http://python.org) (>=3.4) and [Ninja](https://ninja-build.org) (>=1.5).

Before continuing please have a look at the dependencies to guarantee successful compilation.

For each library in the root folder (libmescan/ & libllt/), the following commands have to be executed to fully compile and install linllt. libmescan has to be compiled and installed first.

```bash
meson builddir
cd builddir
ninja
sudo ninja install
```

It might be necessary to call `sudo ldconfig` afterwards, to reload the linker cache.

To use the libraries with Python (>=3.4), the module pylinllt (located in the python_bindings folder) has to be copied to the site-packages of the desired python installation. The module assumes the library in /usr/local/lib.

## Dependencies

Aravis has to be compiled from source via the GNU build system:

- [Aravis](https://github.com/AravisProject/aravis/releases) (version 0.5.8 or newer)

To compile the basic GNU build toolchain and the autotools are necessary. Further aravis dependencies are:

- [libxml2](https://github.com/GNOME/libxml2/releases)
- [glib2](https://github.com/GNOME/glib/releases)

These and the toolchain packages are usually available via the package manager, e.g.

```bash
sudo apt install build-essential autotools-dev automake intltool libtool gtk-doc-tools libxml2-dev libglib2.0-dev
```

Make sure to update and upgrade your repos beforehand. Further information can be found on the [aravis Github page](https://github.com/AravisProject/aravis).

## Using the examples

The C/C++ examples can also be built by meson. For image output support in the C++ container mode example (ContainerMode) [libpng](https://sourceforge.net/projects/libpng/files) (>=1.5) has to be installed.

To use the C example "set_ip_address" and the python script "gige_ip_control.py", make sure that a default gateway route has been added. 

The Python examples need a working installation of matplotlib and scipy.

## Cross compile

To cross compile with the meson build system, different specific cross files can be used. Of course the different cross compilers have to be installed, e.g. for aarch64. The aarch64 environment is set up on the Ubuntu 17.10 distro and can be triggered with:

```bash
meson build --cross-file cross-compilation/aarch64-host.txt
```

At the moment for arm32 compilation a Raspberry Pi is used, thus no cross compilation environment is set up. For this c\_std=c99 (default options in project) and c_args:'-D_POSIX_C_SOURCE=199309L' (shared_library) might be needed in meson.build.

## License

As of version 0.2.x the library is released under LGPLv3. The examples are licensed under MIT license.

ReleaseNotes
v1.0.0 - 20.07.22
-ported to Aravis 0.8

v0.2.6 - 16.03.22
-libmescan updated:
    - compatible with Firmware v52(new Genicam standard)
    - added Sensitivity parameter

v0.2.5 - 08.04.21
-libmescan and pylinllt updated:
    - added 30xx-100
    - Bugfixes

v0.2.4 - 28.05.20
- libmescan and pylinllt updated:
    - Added scanCONTROL 30xx support (200 mm)
    - Bugfixes

v0.2.3 - 21.10.19
- Added scanCONTROL 25xx support (25 mm, 50 mm, 100mm)

v0.2.2 - 14.06.2019
- Updated Documentation

v0.2.1 - 01.12.2018
- Added IP configuration functions
- Added scanCONTROL 30xx support (25 mm, 50 mm)

v0.2.0 - 03.11.2017
- Design: all Macros and typedefs relocated to LLTDataTypes.h
- Design: all functions now in C (libmescan). libllt now C++ wrapper.
- Design: libmescan.h renamed to mescan.h
- Design: libllt.h renamed to llt.h
- Design: device_properties.dat now optional (with file MEDeviceData avail., otherwise scaling and offset only)

- API: removed EventLoop control, use events instead
- API: create Event changed, event handle now return value
- API: WaitForSingleObject timeout now in ms
- API: additional error code and function macros
- API: TransferProfiles parameter "start_stop" changed from bool to gboolean
- API: internal parameters now private and not accessible
- API: added getters for MEDeviceData and ArvDevice
- API: added ExportLLTConfig / ExportLLTConfigString which writes the sensor settings
- API: added ImportLLTConfig / ImportLLTConfigString which reads and sets settings from a string, 
       a .txt file (ExportLLTConfig) or a .sc1 file (Configuration Tools >=5.2)
- API: added functions to read/write post processing arrays
- API: added setter/getter for Ethernet heartbeat timeout
- API: added new feature macros for extra parameter (usable since sensor firmware >v43)
- API: removed ApplicationData. Added getters for scaling and offset instead.
- API: C++ class now wrapper only, renamed to CInterfaceLLT
- API: more entries in TScannerType (respects measuring range)
- API: added SetPeakFilter, SetFreeMeasuringField, SetDynamicMeasuringFieldTracking
- API: added ConvertRearrangedContainer2Values
- API: added API for stream thread control (priority, statistics)
- API: added TranslateErrorValue to convert error codes to readable 
- API: changed get_device_name to comply with Windows API
- API: changed convert_profiles_2_values to comply more with Windows API

- Improved: added polling mode with GetActualProfile
- Improved: invalid Z values now 0 mm
- Bugfix: intensity calculation now correct
- General: general improvements
- General: improved code compatibility
- Added python3 module with bindings (to install copy whole pylinllt folder to pythons "site-packages" 
                                                                                        or "dist-packages" folder)
- Added C and python3 examples
- Added example build files for meson build system
- Note: recommended aravis version: >=0.5.10; minimum for painless operation 0.5.8
  https://github.com/AravisProject/aravis (C)Emmanuel Pacaud <emmanuel@gnome.org>

v0.1.0 - 20.10.2016
- Note: initial release
- Note: supported aravis version: 0.3.7


examples/Python/container_mode.py
#
# scanCONTROL Linux SDK - example code
#
# MIT License
#
# Copyright (c) 2017-2018 Micro-Epsilon Messtechnik GmbH & Co. KG
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:

# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
#
#

import time
import ctypes as ct
import numpy as np
from matplotlib import pyplot as plt
import math as m
import pylinllt as llt

# Parametrize transmission
container_size = 25
scanner_type = ct.c_int(0)

# Init profile buffer and timestamp info
timestamp = (ct.c_ubyte * 16)()
available_resolutions = (ct.c_uint * 4)()
available_interfaces = [ct.create_string_buffer(8) for i in range(6)]
available_interfaces_p = (ct.c_char_p * 6)(*map(ct.addressof, available_interfaces))
lost_profiles = ct.c_uint()

# Create instance and set IP address
hLLT = llt.create_llt_device()

# Get available interfaces
ret = llt.get_device_interfaces(available_interfaces_p, len(available_interfaces))
if ret < 1:
    raise ValueError("Error getting interfaces : " + str(ret))

ret = llt.set_device_interface(hLLT, available_interfaces[0])
if ret < 1:
    raise ValueError("Error setting device interface: " + str(ret))

# Connect
ret = llt.connect(hLLT)
if ret < 1:
    raise ConnectionError("Error connect: " + str(ret))

# Get available resolutions
ret = llt.get_resolutions(hLLT, available_resolutions, len(available_resolutions))
if ret < 1:
    raise ValueError("Error getting resolutions : " + str(ret))

# Set max. resolution
resolution = available_resolutions[0]
ret = llt.set_resolution(hLLT, resolution)
if ret < 1:
    raise ValueError("Error getting resolutions : " + str(ret))

# Declare measuring data arrays
profile_buffer = (ct.c_ubyte * (resolution * 2 * container_size))()
x = (ct.c_double * resolution)()
z = (ct.c_double * resolution)()
intensities = (ct.c_ushort * resolution)()

# Equidistant ranges
x = np.linspace(0, resolution, resolution)
y = np.linspace(0, container_size, container_size)
X, Y = np.meshgrid(x, y)

# Scanner type
ret = llt.get_llt_type(hLLT, ct.byref(scanner_type))
if ret < 1:
    raise ValueError("Error scanner type: " + str(ret))

# Set container to profile config
ret = llt.set_profile_config(hLLT, llt.TProfileConfig.CONTAINER)
if ret < 1:
    raise ValueError("Error setting profile config: " + str(ret))

# Set packet size
ret = llt.set_packet_size(hLLT, 320)
if ret < 1:
    raise ValueError("Error setting packet size: " + str(ret))

# Set trigger
ret = llt.set_feature(hLLT, llt.FEATURE_FUNCTION_TRIGGER, llt.TRIG_INTERNAL)
if ret < 1:
    raise ValueError("Error setting trigger: " + str(ret))

# Set exposure time
ret = llt.set_feature(hLLT, llt.FEATURE_FUNCTION_EXPOSURE_TIME, 100)
if ret < 1:
    raise ValueError("Error setting exposure time: " + str(ret))

# Set idle time
ret = llt.set_feature(hLLT, llt.FEATURE_FUNCTION_IDLE_TIME, 3900)
if ret < 1:
    raise ValueError("Error idle time: " + str(ret))

# Set rearrangement (z only)
rec_log2 = 1.0 / m.log(2.0)
container_resolution = m.floor((m.log(resolution) * rec_log2) + 0.5)
ret = llt.set_feature(hLLT, llt.FEATURE_FUNCTION_PROFILE_REARRANGEMENT, llt.CONTAINER_DATA_Z | llt.CONTAINER_STRIPE_1 | (container_resolution << 12))
if ret < 1:
    raise ValueError("Error setting rearrangement: " + str(ret))

# Set container size
ret = llt.set_profile_container_size(hLLT, 0, container_size)
if ret < 1:
    raise ValueError("Error setting profile container size: " + str(ret))

# Start transfer
ret = llt.transfer_profiles(hLLT, llt.TTransferProfileType.NORMAL_CONTAINER_MODE, 1)
if ret < 1:
    raise ValueError("Error starting transfer profiles: " + str(ret))

# Warm-up time
time.sleep(2)

ret = llt.get_actual_profile(hLLT, profile_buffer, len(profile_buffer), llt.TProfileConfig.CONTAINER,
                           ct.byref(lost_profiles))
if ret != len(profile_buffer):
    raise ValueError("Error get profile buffer data: " + str(ret))

# Stop transmission
ret = llt.transfer_profiles(hLLT, llt.TTransferProfileType.NORMAL_CONTAINER_MODE, 0)
if ret < 1:
    raise ValueError("Error stopping transfer profiles: " + str(ret))

# Disconnect
ret = llt.disconnect(hLLT)
if ret < 1:
    raise ConnectionAbortedError("Error while disconnect: " + str(ret))

ret = llt.del_device(hLLT)
if ret < 1:
    raise ConnectionAbortedError("Error while delete: " + str(ret))

# Convert buffer to big-endian ushort values and reshape them to 2D array
Z = np.frombuffer(profile_buffer, dtype='>H').reshape((container_size, resolution))

fig = plt.figure()
fig.subplots_adjust(wspace=0.3)
plt.pcolormesh(X, Y, Z)
plt.colorbar()
plt.show()


examples/Python/partial_profiles_callback.py
#
# scanCONTROL Linux SDK - example code
#
# MIT License
#
# Copyright (c) 2017-2018 Micro-Epsilon Messtechnik GmbH & Co. KG
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:

# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
#
#

import time
import ctypes as ct
import numpy as np
import threading
import pylinllt as llt

from matplotlib import pyplot as plt
import matplotlib.animation as animation

def profile_callback(data, size, user_data):
    user_data_cast = ct.cast(user_data, ct.POINTER(ct.c_uint))
    if user_data_cast.contents.value == 1:
        global profile_buffer
        ct.memmove(profile_buffer, data, size)
        event.set()

# Parametrize transmission
start_data = 0
data_width = 8
scanner_type = ct.c_int(0)

# Init profile buffer and timestamp info
timestamp = (ct.c_ubyte * 16)()
available_resolutions = (ct.c_uint * 4)()

available_interfaces = [ct.create_string_buffer(8) for i in range(6)]
available_interfaces_p = (ct.c_char_p * 6)(*map(ct.addressof, available_interfaces))

lost_profiles = ct.c_int()
shutter_opened = ct.c_double(0.0)
shutter_closed = ct.c_double(0.0)
profile_count = ct.c_uint(0)
cb_user_data = ct.c_uint(1)  # Sensor 1

get_profile_cb = llt.buffer_cb_func(profile_callback)
event = threading.Event()

# Null pointer if data not necessary
null_ptr_short = ct.POINTER(ct.c_ushort)()
null_ptr_int = ct.POINTER(ct.c_uint)()

hLLT = llt.create_llt_device()

ret = llt.get_device_interfaces(available_interfaces_p, len(available_interfaces))
if ret < 1:
    raise ValueError("Error getting interfaces : " + str(ret))

ret = llt.set_device_interface(hLLT, available_interfaces[0])
if ret < 1:
    raise ValueError("Error setting device interface: " + str(ret))

# Connect
ret = llt.connect(hLLT)
if ret < 1:
    raise ConnectionError("Error connect: " + str(ret))

# Get available resolutions
ret = llt.get_resolutions(hLLT, available_resolutions, len(available_resolutions))
if ret < 1:
    raise ValueError("Error getting resolutions : " + str(ret))

# Set max. resolution
resolution = available_resolutions[0]
ret = llt.set_resolution(hLLT, resolution)
if ret < 1:
    raise ValueError("Error getting resolutions : " + str(ret))

# Declare measuring data arrays
profile_buffer = (ct.c_ubyte*(resolution * data_width))()
x = np.empty(resolution, dtype=float)  # (ct.c_double * resolution)()
z = np.empty(resolution, dtype=float)  # (ct.c_double * resolution)()
x_p = x.ctypes.data_as(ct.POINTER(ct.c_double))
z_p = z.ctypes.data_as(ct.POINTER(ct.c_double))
intensities = (ct.c_ushort * resolution)()

# Partial profile struct
partial_profile_struct = llt.TPartialProfile(0, start_data, resolution, data_width)

# Scanner type
ret = llt.get_llt_type(hLLT, ct.byref(scanner_type))
if ret < 1:
    raise ValueError("Error scanner type: " + str(ret))

# Scanner type
ret = llt.set_resolution(hLLT, resolution)
if ret < 1:
    raise ValueError("Error setting resolution: " + str(ret))

# Set partial profile as profile config
ret = llt.set_profile_config(hLLT, llt.TProfileConfig.PARTIAL_PROFILE)
if ret < 1:
    raise ValueError("Error setting profile config: " + str(ret))

# Set trigger
ret = llt.set_feature(hLLT, llt.FEATURE_FUNCTION_TRIGGER, llt.TRIG_INTERNAL)
if ret < 1:
    raise ValueError("Error setting trigger: " + str(ret))

# Set exposure time
ret = llt.set_feature(hLLT, llt.FEATURE_FUNCTION_EXPOSURE_TIME, 100)
if ret < 1:
    raise ValueError("Error setting exposure time: " + str(ret))

# Set idle time
ret = llt.set_feature(hLLT, llt.FEATURE_FUNCTION_IDLE_TIME, 3900)
if ret < 1:
    raise ValueError("Error idle time: " + str(ret))

# Set partial profile
ret = llt.set_partial_profile(hLLT, ct.byref(partial_profile_struct))
if ret < 1:
    raise ValueError("Error setting partial profile: " + str(ret))

# Register Callback
ret = llt.register_buffer_callback(hLLT, get_profile_cb, ct.byref(cb_user_data))
if ret < 1:
    raise ValueError("Error setting callback: " + str(ret))

# Start transfer
ret = llt.transfer_profiles(hLLT, llt.TTransferProfileType.NORMAL_TRANSFER, 1)
if ret < 1:
    raise ValueError("Error starting transfer profiles: " + str(ret))

# Warm-up time
time.sleep(0.1)

fig, ax = plt.subplots()
line, = ax.plot([], [], ".b", lw=2)
ax.grid()
ax.set_xlim(-60, 60)
ax.set_ylim(25, 450)
line.set_data(x, z)


def data_gen(*args):
    event.wait()
    fret = llt.convert_part_profile_2_values(profile_buffer, len(profile_buffer), ct.byref(partial_profile_struct), scanner_type, 0,
                                         null_ptr_short, null_ptr_short, null_ptr_short, x_p, z_p, null_ptr_int, null_ptr_int)
    if fret & llt.CONVERT_X is 0 or fret & llt.CONVERT_Z is 0:
        raise ValueError("Error converting data: " + str(ret))

    for i in range(16):
        timestamp[i] = profile_buffer[resolution * data_width - 16 + i]

    llt.timestamp_2_time_and_count(timestamp, ct.byref(shutter_opened), ct.byref(shutter_closed), ct.byref(profile_count), null_ptr_short)
    event.clear()

    yield x, z


def update(data):
    ux, uz = data
    line.set_data(ux, uz)
    return line,

ani = animation.FuncAnimation(fig, update, frames=data_gen, interval=40)
plt.show()

ret = llt.transfer_profiles(hLLT, llt.TTransferProfileType.NORMAL_TRANSFER, 0)
if ret < 1:
    raise ValueError("Error stopping transfer profiles: " + str(ret))

# Disconnect
ret = llt.disconnect(hLLT)
if ret < 1:
    raise ConnectionAbortedError("Error while disconnect: " + str(ret))

# Delete
ret = llt.del_device(hLLT)
if ret < 1:
    raise ConnectionAbortedError("Error while delete: " + str(ret))

examples/Python/full_profiles_poll.py
#
# scanCONTROL Linux SDK - example code
#
# MIT License
#
# Copyright (c) 2017-2018 Micro-Epsilon Messtechnik GmbH & Co. KG
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:

# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
#
#

import time
import ctypes as ct
from matplotlib import pyplot as plt
import pylinllt as llt

scanner_type = ct.c_int(0)

# Init profile buffer and timestamp info
timestamp = (ct.c_ubyte*16)()
available_resolutions = (ct.c_uint*4)()
available_interfaces = [ct.create_string_buffer(8) for i in range(6)]
available_interfaces_p = (ct.c_char_p * 6)(*map(ct.addressof, available_interfaces))

lost_profiles = ct.c_uint()
shutter_opened = ct.c_double(0.0)
shutter_closed = ct.c_double(0.0)
profile_count = ct.c_uint(0)

# Null pointer if data not necessary
null_ptr_short = ct.POINTER(ct.c_ushort)()
null_ptr_int = ct.POINTER(ct.c_uint)()

hLLT = llt.create_llt_device()

ret = llt.get_device_interfaces(available_interfaces_p, len(available_interfaces))
if ret < 1:
    raise ValueError("Error getting interfaces : " + str(ret))

ret = llt.set_device_interface(hLLT, available_interfaces[0])
if ret < 1:
    raise ValueError("Error setting device interface: " + str(ret))

# Connect
ret = llt.connect(hLLT)
if ret < 1:
    raise ConnectionError("Error connect: " + str(ret))

# Get available resolutions
ret = llt.get_resolutions(hLLT, available_resolutions, len(available_resolutions))
if ret < 1:
    raise ValueError("Error getting resolutions : " + str(ret))

# Set max. resolution
resolution = available_resolutions[0]
ret = llt.set_resolution(hLLT, resolution)
if ret < 1:
    raise ValueError("Error getting resolutions : " + str(ret))

# Declare measuring data arrays
profile_buffer = (ct.c_ubyte*(resolution*64))()
x = (ct.c_double * resolution)()
z = (ct.c_double * resolution)()
intensities = (ct.c_ushort * resolution)()

# Scanner type
ret = llt.get_llt_type(hLLT, ct.byref(scanner_type))
if ret < 1:
    raise ValueError("Error scanner type: " + str(ret))

# Scanner type
ret = llt.set_resolution(hLLT, resolution)
if ret < 1:
    raise ValueError("Error setting resolution: " + str(ret))

# Set profile config
ret = llt.set_profile_config(hLLT, llt.TProfileConfig.PROFILE)
if ret < 1:
    raise ValueError("Error setting profile config: " + str(ret))

# Set trigger
ret = llt.set_feature(hLLT, llt.FEATURE_FUNCTION_TRIGGER, llt.TRIG_INTERNAL)
if ret < 1:
    raise ValueError("Error setting trigger: " + str(ret))

# Set exposure time
ret = llt.set_feature(hLLT, llt.FEATURE_FUNCTION_EXPOSURE_TIME, 100)
if ret < 1:
    raise ValueError("Error setting exposure time: " + str(ret))

# Set idle time
ret = llt.set_feature(hLLT, llt.FEATURE_FUNCTION_IDLE_TIME, 3900)
if ret < 1:
    raise ValueError("Error idle time: " + str(ret))

# Start transfer
ret = llt.transfer_profiles(hLLT, llt.TTransferProfileType.NORMAL_TRANSFER, 1)
if ret < 1:
    raise ValueError("Error starting transfer profiles: " + str(ret))

# Warm-up time
time.sleep(0.2)

ret = llt.get_actual_profile(hLLT, profile_buffer, len(profile_buffer), llt.TProfileConfig.PROFILE,
                           ct.byref(lost_profiles))
if ret != len(profile_buffer):
    raise ValueError("Error get profile buffer data: " + str(ret))

ret = llt.convert_profile_2_values(profile_buffer, len(profile_buffer), resolution, llt.TProfileConfig.PROFILE, scanner_type, 0,
                                null_ptr_short, intensities, null_ptr_short, x, z, null_ptr_int, null_ptr_int)
if ret & llt.CONVERT_X is 0 or ret & llt.CONVERT_Z is 0 or ret & llt.CONVERT_MAXIMUM is 0:
    raise ValueError("Error converting data: " + str(ret))

for i in range(16):
    timestamp[i] = profile_buffer[resolution * 64 - 16 + i]

llt.timestamp_2_time_and_count(timestamp, ct.byref(shutter_opened), ct.byref(shutter_closed), ct.byref(profile_count), null_ptr_short)

# Stop transmission
ret = llt.transfer_profiles(hLLT, llt.TTransferProfileType.NORMAL_TRANSFER, 0)
if ret < 1:
    raise ValueError("Error stopping transfer profiles: " + str(ret))

# Disconnect
ret = llt.disconnect(hLLT)
if ret < 1:
    raise ConnectionAbortedError("Error while disconnect: " + str(ret))

# Delete
ret = llt.del_device(hLLT)
if ret < 1:
    raise ConnectionAbortedError("Error while delete: " + str(ret))

plt.figure(facecolor='white')
plt.subplot(211)
plt.grid()
plt.xlabel('x')
plt.ylabel('z')
plt.xlim(-60, 60)
plt.ylim(25, 450)
plt.plot(x, z, 'g.', label="z", lw=2)
plt.subplot(212)
plt.grid()
plt.xlabel('x')
plt.ylabel('intensities')
plt.xlim(-30, 30)
plt.ylim(0, 1200)
plt.plot(x, intensities, 'r.', label="intensities", lw=1)
plt.show()